### ISSUE TITLE
Link states together

### ISSUE THREAD

**Issue by galgalesh** [juju-solutions/charms.reactive#97]
This might be a more permanent solution to the issues @stub42 is facing with the apt layer. A simplification of the issue in question:

```python
@when("config.changed.packages")
@when("apt.sources.configured")
def  install_packages():
    # ...

@when("config.changed.sources")
def configure_sources():
    # ...
```

The issue here is that `configure_sources` has to be run before `install_packages` if both `config.changed.packages` and `config.changed.sources` are set.

I'd like to find a solution that isn't tied to implementation details of the reactive framework (ie. the fact that a "run" starts as part of a hook). Nor to the technical details of this issue. An [`@preflight` decorator](https://github.com/juju-solutions/charms.reactive/pull/20) solves this issue because config cannot change during an reactive run. However, other stuff can change such as files changing or states being added and removed. We also want to cover those cases.

I don't think the reactive framework can currently solve this issue in a clean way.

A solution for this problem that fits the reactive approach better is to be able to link states together so that state changes can trigger other state changes without the need for a handler to run.

```python
# Every time `config.changed.sources` gets set, then `apt.sources.configured` gets removed
# immediately; before running the next handler.
reactive.if_this_then_remove('config.changed.sources', 'apt.sources.configured')

@when("config.changed.packages")
@when("apt.sources.configured")
def  install_packages():
    # ...

@when("config.changed.sources")
def configure_sources():
    # ...
    set_state('apt.sources.configured')
```

I think that solves @stub42's issue. Thoughts?

**Comment by johnsca** [comment]
I think this can actually be solved purely using reactive states / flags:

```python
@when("config.changed.packages")
@when("apt.sources.configured")
@when_not("config.changed.sources")
def  install_packages():
    # ...

@when("config.changed.sources")
def configure_sources():
    # ...
    remove_state("config.changed.sources")
    set_state("apt.sources.configured")
```

This makes sense to me, as it's essentially saying, "We can install packages as long as we don't have any unprocessed changes to the sources list."

**Comment by galgalesh** [comment]
Yes, that's a lot better :)

**Comment by stub42** [comment]
Sleeping on the issue, there is a flaw in the plan and I think I'm still stuck needing a `@preflight` equivalent.

Even if the apt layer sets a flag when the system is ready, that only helps for code explicitly using the apt layer. The snap layer, for a real example, does an explicit 'apt install snapd' if snapd is not available (so it works under trusty). It does not use the apt layer, because the snap layer needs to work on snap-only systems. It cannot cannot wait for the apt layer to set a flag, because the charm may not be using the apt layer. But this affects charms using both the apt and snap layers, as they may end up running apt commands before the apt layer has been initialized.

The same issue exists if the apt layer reconfigures the apt subsystem to set reactive states automatically - other layers may attempt to run apt commands before the apt layer has initialized things so they actually work in the deployment environment.

I think this is the same underlying issue that the leadership layer had, except in that case we thought it appropriate to move the leadership layer into core.

And now I think of it, the coordinator layer has the same problem. It implements distributed locking using leadership and the peer relation, so you can easily do things like rolling restarts. If it does not have an opportunity to run code before the queued reactive handlers are run, then handlers may run with expired locks.

**Comment by galgalesh** [comment]
The more I think of it, the more I notice that the `@preflight` hook won't solve all possible occurrences of this issue. Let's look at the example of a Puppet layer. Puppet modules have to be installed before a higher layer can run `puppet apply`. At first sight, the `@preflight` hook makes this possible. However, the Puppet layer uses the apt layer to install Puppet itself. So the order we want is 

1. Apt layer configures sources
2. Apt layer installs puppet
3. Puppet layer installs sources
4. Puppet layer installs modules
5. top-level layer runs `puppet apply`.

In order to achieve this, we would need the `@preflight` handlers to run in a specific order.

About your issues:

- In the snap layer example, you're basically re-implementing the apt layer, so the solution for the apt layer should work for the snap layer too.
- Charms that use the apt layer and run apt commands without using your api are broken. If the api doesn't provide for all use-cases then we should look at fixing the api. I'd be happy to take a look at the code to figure this out. Maybe we can do a small sprint at the Summit?
- Cory's proposal for adding objects to flags would allow us to create a much better api for the apt layer that eliminates the need for any upper layer to either call apt manually or use the python apt library.

I think it's clear that this kind of stuff is hard, but the `@preflight` handler seems to be a hack that is tied very closely to the technical details of your issue.

I think that in general, the issue is that we don't have a way to instantly invalidate flags when something happens. `apt.installed.x` has to be unset immediately when landscape removes x. `apt.sources.configured` has to be unset immediately when the sources change.

**Comment by stub42** [comment]
On 1 February 2017 at 17:11, Merlijn Sebrechts <notifications@github.com>
wrote:

> The more I think of it, the more I notice that the @preflight hook won't
> solve all possible occurrences of this issue. Let's look at the example of
> a Puppet layer. Puppet modules have to be installed before a higher layer
> can run puppet apply. At first sight, the @preflight hook makes this
> possible. However, the Puppet layer uses the apt layer to install Puppet
> itself. So the order we want is
>
>    1. Apt layer configures sources
>    2. Apt layer installs puppet
>    3. Puppet layer installs sources
>    4. Puppet layer installs modules
>    5. top-level layer runs puppet apply.
>
> In order to achieve this, we would need the @preflight handlers to run in
> a specific order.
>
This example only works if, for some reason, the Puppet layer needs to make
use of the apt layer during the setup phase (before the queue of handlers
gets processed). What is much more likely is that it uses
@when('apt.installed.puppet') like any other consumer of the apt layer.

I have not seen this sort of dependency problem. If it did happen, yes,
@preflight cannot help with ordering and the maintainer of the Pupper layer
would need to coordinate with me to work out some other solution. I do not
think we should add any sort of dependency chain to @preflight handlers; if
we need it, I'd instead propose adding a 'priority' to handlers and turning
the arbitrary ordered queue of handlers into a priority queue.



> About your issues:
>
>    - In the snap layer example, you're basically re-implementing the apt
>    layer, so the solution for the apt layer should work for the snap layer too.
>    - Charms that use the apt layer and run apt commands without using
>    your api are broken. If the api doesn't provide for all use-cases then we
>    should look at fixing the api. I'd be happy to take a look at the code to
>    figure this out. Maybe we can do a small sprint at the Summit?
>    - Cory's proposal for adding objects to flags would allow us to create
>    a much better api for the apt layer that eliminates the need for any upper
>    layer to either call apt manually or use the python apt library.
>
> I now think my example with the snap layer is poor and I can't defend it.
The snap layer doesn't use the apt layer because it doesn't want the apt
specific configuration in config.yaml, and because it doesn't want or need
it then I cannot argue that the apt layer needs to make use of this
(non-existent) configuration during a preflight stage.

I had not thought about attaching package data to flags. Yes, there are
some interesting ideas there. Should handlers be reinvoked if the data
attached to their flags changes? Or are they only invoked when the flag is
first raised?


>
>
> I think it's clear that this kind of stuff is hard, but the @preflight
> handler seems to be a hack that is tied very closely to the technical
> details of your issue.
>
> I think that in general, the issue is that we don't have a way to
> instantly invalidate flags when something happens. apt.installed.x has to
> be unset immediately when landscape removes x. apt.sources.configured has
> to be unset immediately when the sources change.
>

I would not say that the @preflight handler is a hack tied very closely to
the technical details of my issue. It is a tool that neatly and easily let
me solve the technical details of several different issues.

Yes, I think in general it comes down to a way to invalidate flags when
something happens. I think @preflight would be a general way to do this,
providing an opportunity to do this between 'something happens' and the
queue of handlers being processed. It does not account for things happening
entirely outside the hook context, so if a cronjob for instance does a
'charms.reactive get_states' then no handlers (including @preflight) would
kick in. The original proposal in this thread actually could work in this
case, as the rules could be processed without needing the overhead of
starting a hook context (and blocking if something else is running in a
hook context).


-- 
Stuart Bishop <stuart.bishop@canonical.com>

**Comment by ajkavanagh** [comment]
W.r.t the suggestion in this issue:

```python
# Every time `config.changed.sources` gets set, then `apt.sources.configured` gets removed
# immediately; before running the next handler.

reactive.if_this_then_remove('config.changed.sources', 'apt.sources.configured')
```

My opinion is that this probably isn't a good idea.  My main objections are:

1. It's file/module level code that runs during the loading of the module.  I think that this is a _poor_ place to put code that is actually the functionality of the module, rather than the setting up of the module. 
2. Handlers are the place to switch on/off states (flags).  This is already a pattern that has been established, and this change would further complicate the set of concepts that a charm author would need to know about.

However, states (flags) are the principal (only?) method of communicating between layers/interfaces without using a separate class cuts across them - which itself further complicates charm authoring, as I think that it runs roughshod over the intentions of layers in charms.reactive.

There does seem to be a genuine need for `@preflight` but it feels like something that should be genuinely discouraged for general use as again, adding ordering would further complicate the system, and without ordering,*too* many `@preflight` decorators would become unmanageable.

The same is largely true of an `@atexit` type decorator for some of the same reasons.

Just my 2c as they say; hopefully something useful in there.

**Comment by galgalesh** [comment]
> 1. It's file/module level code that runs during the loading of the module. I think that this is a poor place to put code that is actually the functionality of the module, rather than the setting up of the module.

I agree, my example is mostly to illustrate the concept. The actual implementation should address this issue.

> 2. Handlers are the place to switch on/off states (flags). This is already a pattern that has been established, and this change would further complicate the set of concepts that a charm author would need to know about.

I'd argue that `@preflight` actually complicates the concepts. To use `@preflight`, you need to know 
1. how flags work,
2. how the reactive framework works internally,
3. how hooks work.

Ideally, users should only have to know about 1. so the more we solve with the "flags" concept, the easier the reactive framework is to use. This complexity has the potential to leak to higher layers, it isn't limited to the developer of the layer that actually uses `@preflight`.

Furthermore, what we try to create with `@preflight` are flags that magically appear or disappear outside of the "normal" functionality of the reactive framework. With `if_this_then_remove`, we make these changes explicit.

Lastly, the `if_this_then_remove` proposal doesn't only solve the `@preflight` use-case. It also enables developers to create layers that validate config. See issue https://github.com/juju-solutions/charms.reactive/issues/98

**Comment by preflight** [comment]
@preflight here - think I might need to change my username :-)

**Comment by ajkavanagh** [comment]
@preflight lol - sorry! I'll try to remember to refer to it as the `preflight` decorator!

**Comment by johnsca** [comment]
The fundamental issue that I have with `@preflight` is that the timing of it is relative to something conceptually outside of the reactive framework but it is expected to block processing of handlers while it runs.  The concept of "mutually exclusive flags," or even "pre-set callbacks" on flags fits better with the framework, in that sense, to me.  However, both end up interrupting the conceptual flow of "when a flag is changed, the handlers that care about it are queued to run," so they both still make me a bit uneasy.  But, obviously there's a definite use-case for it, so my preference would be for the descriptive approach ("these flags are mutually exclusive").  The flag classes should give a good place to implement the behavior without it being at the module level.

> Should handlers be reinvoked if the data attached to their flags changes?

I definitely think we'd need to be able to mark flags as "changed" based on data or some other condition, even if they're already set.

**Comment by galgalesh** [comment]
> The fundamental issue that I have with @preflight is that the timing of it is relative to something conceptually outside of the reactive framework but it is expected to block processing of handlers while it runs.

Yes, in OOP terms, the underlying hook mechanics leak into the reactive layers. We'd rather not have those leaks because of complexity and because we'd like to get the reactive framework in places where hooks don't exist.

> The concept of "mutually exclusive flags," or even "pre-set callbacks" on flags fits better with the framework, in that sense, to me. However, both end up interrupting the conceptual flow of "when a flag is changed, the handlers that care about it are queued to run,"

I'm not sure what the right concept is. Is it possible that the queue is just not the correct concept here? Alex had a suggestion about this, and although I don't completely agree with his proposal, he might be right that the queue is the conceptual issue. I don't agree because his solution would prevent handlers from running multiple times. I think this is necessary and even wanted in some cases, like in the case of an update status hook. My proposal and more info about this at the end of this comment.

> so they both still make me a bit uneasy. But, obviously there's a definite use-case for it, so my preference would be for the descriptive approach ("these flags are mutually exclusive"). The flag classes should give a good place to implement the behavior without it being at the module level.


>
> > Should handlers be reinvoked if the data attached to their flags changes?
>
> I definitely think we'd need to be able to mark flags as "changed" based on data or some other condition, even if they're already set.
>

Can we treat "the flag data changes" exactly like "this flag was unset and set again"? That seems logical to me..


# [proposal] Two sets instead of queue

You have two sets of handlers: `dormant` and `handled`. All handlers start in the `dormant` set.

1. Test each dormant handler and execute the first one that matches.
2. Put the handler in the `handled` set.
3. Check the `handled` set for handlers whose preconditions are false.
4. Go back to 1, start at *the beginning* of the `dormant` set.

What we get with this:

 - We get control over the order that handlers are executed.
   - We can eliminate the *"`restart` handler causes multiple restarts"* issue by putting the restart handler at the end of the file.
   - We get the wanted behavior of *"the `status_set` handler gets executed each time the status changes"* by putting that handler first.
- The order in which the handlers are executed will always be the same -> reproducible.
- Handlers are sure that their preconditions are met when they are executed. -> Less checking of state necessary.

The order in the dormant set is the order in the layers and files. The highest handler in the base layer is the first handler in the `dormant` set. The lowest handler in the top-level layer is the last handler in the `dormant` set. This would mean that we can put a `status_set` handler in the base layer that is executed each time when any layer changes the status (changes the object of the status flag).


Thoughts?

*PS: I don't think that this is harder to understand than what the current concept is, especially since this proposal doesn't leak the "hooks" concept at all. Handlers aren't re-executed in a following hook unless they get thrown out of the "handled" set.*

**Comment by ajkavanagh** [comment]
@galgalesh This is interesting. One worry I have is around compatibility between charms.reactive 1 and 2, but I'll write that up (hopefully) in another issue.

The thing I like from what @galgalesh proposes is that the conditions stated on the handler are still true when the handler is eventually run.  This, I feel, is important from a charm author perspective as I think it would be a surprise if this wasn't the case (i.e. the conditions are written above the function)!

I don't think we should enforce order of handler execution as that would be difficult to determine in a charm with several of layers.  It might be implementation dependent on the order in which reactive walks the filesystem, and thus imports the modules, and hence the order might be arbitrary?  Thus, we probably shouldn't make this a pre-condition for this feature/topic.

### A slightly different take on the proposal (this is still a work in progress)

So I'd like to declare the queue dead too.  (Sort of).  Instead we might have:

* A list of all possible handlers
* A list of handlers that have been run

Then:

1. Evaluate the conditions for the first handler in the list.  If true, then run the handlers and then add it to the handlers that have been run list.
2. Continue until the end of the list.
3. Start back at 1 with the remaining handlers that haven't run.
4. Exit if no handlers were run during a pass through the remaining list.

Consequences:

a. A handler will at most be run only once.
b. The conditions on the handler will always be true when the handler is run.

Issues:

a. A handler will at most be run only once.
b. The order is indeterminate.
c. There is a hard to debug bug lurking.

The bug is:

```python
@when('flag1')
def funca():
    remove_state('flag1')

@when('flag1')
def funcb():
    # do something
```

i.e. depending on the order of the functions, 'do something' may or may not happen.  I think that this is a error in writing the charm.  Perhaps we can flag this during charm build to indicate that two functions have exactly the same conditions?

I think there are other issues with this, but I wanted to add it to the discussion to see if it provokes further thought in the feature.,
